lib/code_corps/github.ex

```diff
@@ -1,41 +1,273 @@
+        message: message,
+        status_code: status_code
+      }
+    end
+
+    @spec new({integer, map}) :: t
+    def new({status_code, %{"message" => message, "documentation_url" => documentation_url}}) do
+      %__MODULE__{
+        documentation_url: documentation_url,
+        message: message,
+        status_code: status_code
+      }
+    end
+
+    @spec new({integer, map}) :: t
+    def new({status_code, %{"message" => message}}) do
+      %__MODULE__{
+        message: message,
+        status_code: status_code
+      }
+    end
+
+    defp convert_error(%{"code" => code, "field" => field, "resource" => resource}) do
+      APIErrorObject.new([code: code, field: field, resource: resource])
+    end
+  end
+
+  defmodule HTTPClientFailed do
+    defstruct [:reason, message: """
+    The GitHub HTTP client encountered an error while communicating with
+    the GitHub API.
+    """]
+
+    def new(opts) do
+      struct(__MODULE__, opts)
+    end
+  end
+
+  @type method :: :get | :post | :put | :delete | :patch
+  @type headers :: %{String.t => String.t} | %{}
+  @type body :: {:multipart, list} | map
+  @typep http_success :: {:ok, integer, [{String.t, String.t}], String.t}
+  @typep http_failure :: {:error, term}
+
+  @type api_error_struct ::
+    %APIError{}            |
+    %HTTPClientFailed{}
+
+  @spec get_base_url() :: String.t
+  defp get_base_url() do
+    Application.get_env(:code_corps, :github_base_url) || "https://api.github.com/"
+  end
+
+  #
+  @spec get_token_url() :: String.t
+  defp get_token_url() do
+    Application.get_env(:code_corps, :github_oauth_url) || "https://github.com/login/oauth/access_token"
+  end
+
+  @spec use_pool?() :: boolean
+  defp use_pool?() do
+    Application.get_env(:code_corps, :github_api_use_connection_pool)
+  end
+
+  @spec add_default_headers(headers) :: headers
+  defp add_default_headers(existing_headers) do
+    Map.merge(%{"Accept" => "application/vnd.github.machine-man-preview+json"}, existing_headers)
+  end
+
+  @spec add_access_token_header(headers, String.t | nil) :: headers
+  defp add_access_token_header(existing_headers, nil), do: existing_headers
+  defp add_access_token_header(existing_headers, access_token) do
+    Map.put(existing_headers, "Authorization", "token #{access_token}")
+  end
+
+  def generate_jwt do
+    # generates JWT with RSA key, using 10 minute expires time, RS256 algo,
+    # and issuer of the GitHub app's id
+    %{}
+    |> Joken.token
+    |> Joken.with_exp(Timex.now |> Timex.shift(minutes: 10) |> Timex.to_unix)
+    # replace this one with integration identifier
+    # https://developer.github.com/apps/building-integrations/setting-up-and-registering-github-apps/about-authentication-options-for-github-apps/#authenticating-as-a-github-app
+    |> Joken.with_iss("foo")
+    |> Joken.with_iat(Timex.now |> Timex.to_unix)
+    |> Joken.with_signer(Joken.rs256(rsa_key))
+    |> Joken.sign
+    |> Joken.get_compact
+  end
```

Comment:
This is almost done. What's left is to 

- [get a private key file](https://developer.github.com/apps/building-integrations/setting-up-and-registering-github-apps/registering-github-apps#generating-a-private-key)
- [get the app id](https://developer.github.com/apps/building-integrations/setting-up-and-registering-github-apps/about-authentication-options-for-github-apps/#authenticating-as-a-github-app)